kernel/mutex.c

```diff
@@ -247,6 +239,9 @@ void _impl_k_mutex_unlock(struct k_mutex *mutex)
 		mutex->owner = new_owner;
 		mutex->lock_count++;
 		mutex->owner_orig_prio = new_owner->base.prio;
+
+		_ready_thread(new_owner);
+		irq_unlock(key);
 	} else {
 		irq_unlock(key);
```

Comment:
Maybe you can move the call to irq_unlock() outside this if/else block, right before k_sched_unlock()?

(I'm not sure if there's anything that assumes that mutex->owner will be set to NULL with the interrupts unlocked, though.)

Reply:
Yeah, that's cleaner. Honestly I think putting the owner assignment outside the lock is a race to begin with. And in fact the block here is tested by if(new_owner) and mutex->owner gets set to that, so we can move the asssignment to before the block and the unlock to after and it's cleaner still. Thanks.